### PR TITLE
logging TA grading

### PR DESCRIPTION
**Pull request by saxocellphone**
This will make a new directory in the /logs folder called ta_grading. Will log whenever grader open grading page, open component, and save component.

### CODE REVIEW

File: site/app/libraries/Logger.php [hunk 1]
```diff
@@ -198,4 +198,33 @@ public static function logAccess($user_id, $token, $action) {
...
     }
+
+
+    /**
+     * This logs the grading activity of any graders when they
+     * 1. Open the student's page to grade
+     * 2. Opening a component
+     * 3. Saving a component
+     * The log is in the format of
+     * Timestamp | Gradeable_id | Grader ID | Student ID | Component_ID (except for case 1) | Action | User Agent
```
Comment (KevinMackenzie) [inline site/app/libraries/Logger.php]: we should just have a default value so every row as the same number of columns (like id=-1)

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 2]
```diff
@@ -1115,6 +1116,15 @@ public function showGrading() {
         $late_status = $old_gradeable->calculateLateStatus();
         // TODO: End region
 
+        $logger_params = array(
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "submitter_id" => $submitter_id,
+            "action" => "VIEW_PAGE",
+            "token" => $_COOKIE['submitty_token']
```
Comment (KevinMackenzie) [inline site/app/controllers/grading/ElectronicGraderController.php]: remove the token here...

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 3]
```diff
@@ -1373,6 +1383,16 @@ public function ajaxSaveGradedComponent() {
             $silent_edit = false;
         }
 
+        $logger_params = array(
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "component_id" => $component_id,
+            "action" => "SAVE_COMPONENT",
+            "token" => $_COOKIE['submitty_token'],
```
Comment (KevinMackenzie) [inline site/app/controllers/grading/ElectronicGraderController.php]: ...and here ...

File: site/app/controllers/grading/ElectronicGraderController.php [hunk 4]
```diff
@@ -2109,6 +2129,17 @@ protected function ajaxGetGradedComponent() {
         // Get / create the graded component
         $graded_component = $ta_graded_gradeable->getGradedComponent($component, $grader);
 
+        $logger_params = array(
+            "gradeable_id" => $gradeable_id,
+            "grader_id" => $this->core->getUser()->getId(),
+            "component_id" => $component_id,
+            "action" => "OPEN_COMPONENT",
+            "token" => $_COOKIE['submitty_token'],
```
Comment (KevinMackenzie) [inline site/app/controllers/grading/ElectronicGraderController.php]: ...and here

File: .setup/install_system.sh [hunk 5]
```diff
@@ -640,10 +640,13 @@ if [ ${WORKER} == 0 ]; then
         mkdir -p ${SUBMITTY_REPOSITORY}/.vagrant/${DISTRO}/${VERSION}/logs/submitty/site_errors
         ln -s ${SUBMITTY_REPOSITORY}/.vagrant/${DISTRO}/${VERSION}/logs/submitty/access ${SUBMITTY_DATA_DIR}/logs/access
         ln -s ${SUBMITTY_REPOSITORY}/.vagrant/${DISTRO}/${VERSION}/logs/submitty/site_errors ${SUBMITTY_DATA_DIR}/logs/site_errors
+        ln -s ${SUBMITTY_REPOSITORY}/.vagrant/${DISTRO}/${VERSION}/logs/submitty/site_errors ${SUBMITTY_DATA_DIR}/logs/ta_grading
```
Comment (MasterOdin) [inline .setup/install_system.sh]: Change `site_errors` on this line to `ta_grading`.

File: site/app/libraries/Logger.php [hunk 6]
```diff
@@ -198,4 +198,31 @@ public static function logAccess($user_id, $token, $action) {
...
+        $log_message[] = static::getTimestamp();
+        $log_message[] = $params['gradeable_id'];
+        $log_message[] = $params['grader_id'];
+        $log_message[] = $params['submitter_id'];
+        $log_message[] = array_key_exists('component_id', $params) ? $params['component_id'] : "-1";
+        $log_message[] = $params['action'];
+        $log_message[] = $_SERVER['HTTP_USER_AGENT'];
+        $log_message = implode(" | ", $log_message)."\n";
+        @file_put_contents(FileUtils::joinPaths(static::$log_path, 'ta_grading', "{$filename}.log"), $log_message, FILE_APPEND | LOCK_EX);
+    }
```
Comment (MasterOdin) [inline site/app/libraries/Logger.php]: Let's combine this function with the one above such that both call a new function:

```
private static function logMessage($folder, $log_message) {
    $filename = static::getFilename();
    array_unshift($log_message, static::getTimestamp());
    $log_message[] = $_SERVER['HTTP_USER_AGENT'];
    $log_message = implode(" | ", $log_message)."\n";
    @file_put_contents(FileUtils::joinPaths(static::$log_path, 'ta_grading', "{$filename}.log"), $log_message, FILE_APPEND | LOCK_EX);
}
```

and thus this function would be:
```
    public static function logTAGrading($params){
        $log_message[] = $params['gradeable_id'];
        $log_message[] = $params['grader_id'];
        $log_message[] = $params['submitter_id'];
        $log_message[] = array_key_exists('component_id', $params) ? $params['component_id'] : "-1";
        $log_message[] = $params['action'];
        $log_message[] = $_SERVER['HTTP_USER_AGENT'];
        static::logMessage('ta_grading', $log_message);
    }
```

**Comment by saxocellphone** [comment]
I'm not sure why the Unit tests are failing, I made sure the dir in LoggerTester is the ta_grading folder, but apparently the test can't find it.

File: site/tests/app/libraries/LoggerTester.php [hunk 7]
```diff
@@ -104,8 +103,34 @@ public function testInvalidDirectory() {
...
+        $datetime = explode(" ", $line[0]);
+        $time = explode(":", $datetime[0]);
+        $date = explode("/", $datetime[1]);
+        $this->assertTimeEqualsCurrent($date, $time);
+        $this->assertEquals(2, strlen($time[2]));
+        $this->assertEquals("test_gradeable", $line[1]);
+        $this->assertEquals("instructor", $line[2]);
+        $this->assertEquals("aphacker", $line[3]);
+        $this->assertEquals("1", $line[4]);
+        $this->assertEquals("VIEW_PAGE", $line[5]);
```
Comment (MasterOdin) [inline site/tests/app/libraries/LoggerTester.php]: It looks like these asserts should be (based on the `$logging_params`):
```
        $this->assertEquals("test_grader", $line[2]);
        $this->assertEquals("test_submitter", $line[3]);
        $this->assertEquals("1", $line[4]);
        $this->assertEquals("test_action", $line[5]);
```

Comment (saxocellphone) [reply site/tests/app/libraries/LoggerTester.php]: Oh my goodness I keep changing one line of code each time thinking that was the only mistake. Sorry this took so many tries

**Comment by bmcutler** [comment]
@saxocellphone 
Please add semester and course name to the ta grading logging
thanks!

**Pull request merged**